[ add ] `Bool` action on a `RawMonoid` plus properties

jamesmckinna opened the pull request:
Prompted by @JacquesCarette 's desire to further systematise the refactoring in #2407 , this adds two operations which appear *ad hoc* in the normal form semantics for `IdempotentCommutativeMonoid`s, but are yet more instances of a `(Raw)MonoidAction` cf. #2348 / #2350 by analogy with the corresponding actions in the normal form semantics for `CommutativeMonoid`s. 

NB 
* Nat-action arises because (Nat, _+_, 0) is the free commutative monoid on one generator...
* similarly, (Bool, /\, true) is, I think, the free idempotent commutative monoid on one generator?

Could be left as DRAFT, pending further work cf. #2351 , but worth inviting review now, I think.

NB. possible issues:
* names of things: current notation is *terrible*, unfortunately, so suggestions welcome!
* left-/right- symmetric versions, ie. `x + n × y` as well as `n × x + y` etc.
* possible refactoring of the existing definitions as instances of these generalised 'affine' ones?
* refactoring not yet applied to `Algebra.Solver.*Monoid` etc.
* ... further systematisation of existing `Algebra.Definitions.RawMonoid._×_` operation etc. possible as well?
* Fairbairn threshhold?

jamesmckinna commented:
Ooops. Moving to DRAFT while I rethink this.

gallais commented:
Conor's disembodied voice tells me we should use a non-symmetric infix symbol for this non-symmetric action.

jamesmckinna commented:
> Conor's disembodied voice tells me we should use a non-symmetric infix symbol for this non-symmetric action.

I'm genuinely torn over this. I understand, appreciate, and even largely agree with, the doctrine (prevalent in WG2.1) that non-symmetric/commutative operations should be given non-symmetric notation. Nevertheless, we don't insist that that a `Group` operation be given a non-symmetric symbol... even if we *do* distinguish it as `_+_` (or something else 'obviously connoting commutativity') in the `AbelianGroup` case.

Nevertheless, I think we do ourselves (or else: type theory as a principled metalanguage for the prosecution of such doctrine) a disservice by taking the 'old-fashioned' CFG view of notation not coming *a priori* with a type attached.

The case here, as more generally in the [draft PR](https://github.com/agda/agda-stdlib/pull/2350) on `Algebra.Action`, is that the point of an action is that it is *extending* the domain of applicability of a given operation (however symbolised) to the structure on which it acts... and the generalised associativity property of a `MagmaAction` is *precisely* where the extended notation bites, visually/mnemonically at least.

You might object that the `Bool` action defined here, or `Algebra.Definitions.RawMonoid. _×_` aren't in and of themselves 'commutative'... but that obscures the fact that their `flip`ped counterparts are in fact extensionally equal to these definitions.

So perhaps the way out of the bottle is to tag the symbol with `ˡ` and `ʳ` annotations, and use infix symbol positions to record where the tail-recursive accumulator argument should go in the extended notation, yielding four compound symbols in all?

I guess I could live with that... modulo leaving `_×_` unchanged as 'legacy' (yuk)?

MatthewDaggitt commented on src/Algebra/Properties/Monoid/Mult.agda:
```diff
@@ -78,3 +79,16 @@ open import Algebra.Definitions.RawMonoid rawMonoid public
   n × x + (m ℕ.* n) × x ≈⟨ ×-homo-+ x n (m ℕ.* n) ⟨
   (suc m ℕ.* n) × x     ∎
+
+-- _∧_ is homomorphic with respect to _Bool∧_.
```
The comment should be `Bool._∧_`

MatthewDaggitt commented on src/Algebra/Properties/Monoid/Mult.agda:
```diff
@@ -78,3 +79,16 @@ open import Algebra.Definitions.RawMonoid rawMonoid public
+∧-assocˡ false b x = refl
+∧-assocˡ true  b x = refl
+
+∧∙-≗∧+ : ∀ b x y → b ∧′ x ∙ y ≈ (b ∧ x) + y
```
This is a very incomprehensible name. Maybe `b∧x∙y≈b∧x+y` would be better?

MatthewDaggitt reviewed (commented):
Otherwise looks good.

JacquesCarette commented on src/Algebra/Definitions/RawMonoid.agda:
```diff
@@ -65,3 +66,18 @@ suc n ×′ x = n ×′ x + x
+infixr 8 _∧_
+
+_∧_ : Bool → Carrier → Carrier
+b ∧ x = if b then x else 0#
```
Strictly speaking, this could be defined for any `Pointed` structure with an apartness relation, as an action on a `Pointed` structure.

How is the above 'conjuction'?

jamesmckinna replied:
OK, I think that's fair... I was being pretty *ad hoc* in wanting to add this by analogy with the `Nat` action already defined there, as then these two instances of a phenomenon (eventually: wreath product) are the relevant ones for the `Algebra.Solver.*CommutativeMonoid` refactorings #2407 #2457 ... and I'm still unhappy about the general picture for `Pointed` things in the library, so don't (yet) want to reopen that can of worms...

As for it being 'conjunction', well it clearly generalises it: for `Carrier` being that of the `Monoid` structure on `Bool` given by `_∨_` with unit `false`, 
```agda
b ∧ x = if b then x else false
```
is a definition of 'conjunction'... yes!?

JacquesCarette replied:
Yes, that is the operational definition of 'conjunction' on `Bool`. 

I guess that I was thinking along the lines of "if b is the least element then x else some-distinguished-point" as the general meaning. It feels more action-y. 'conjunction' feels like it's a coincidence when `Carrier` is `Bool`.

JacquesCarette commented on src/Algebra/Definitions/RawMonoid.agda:
```diff
@@ -65,3 +66,18 @@ suc n ×′ x = n ×′ x + x
+_∧_ : Bool → Carrier → Carrier
+b ∧ x = if b then x else 0#
+
+-- tail-recursive optimisation
```
optimisation of *what*? It doesn't seem like an optimisation of the above? What am I missing?

jamesmckinna replied:
As with the TCOptimised forms of (monoid) sum and multiplication, the 'optimisation' lies in avoiding having to explicitly track terms of the form `x + 0#` when these can be computed away instead to `x`... But I admit I haven't *quite* got my story straight yet... ;-)

jamesmckinna replied:
As witnessed by `b∧x∙y≈b∧x+y`...

JacquesCarette replied:
It still feels like a stretch. I do like the optimization of an action not doing a (statically known) null operation, a lot. I'm still worried that this particular case is a coincidence rather than fundamental!

jamesmckinna replied:
Well I think it extends to the `Nat` cases as well (but obscured in the current presentation, which doesn't introduce the tail-recursive forms...)

JacquesCarette commented on src/Algebra/Properties/Monoid/Mult.agda:
```diff
@@ -78,3 +79,16 @@ open import Algebra.Definitions.RawMonoid rawMonoid public
+-- _∧_ is homomorphic with respect to _Bool∧_.
+
+∧-homo-true : ∀ x → true ∧ x ≈ x
+∧-homo-true x = refl
```
Here and below: I quite like the use of `_` for arguments that you're never going to look at in the rhs.

JacquesCarette commented:
What prevents this from losing its DRAFT status?

jamesmckinna commented:
> What prevents this from losing its DRAFT status?

Off the top of my head:
* no bandwidth to think about the problem... incl revisiting
* does it solve the problem I wanted it to?
* does it help with #2457 as I had hoped?

All of the above are slightly beyond my capacity to think right now, so `DRAFT` seemed to safest way to mark "I'll come back to this (at some point)"

gallais commented on CHANGELOG.md:
````diff
@@ -48,6 +48,19 @@ New modules
 
+* In `Algebra.Definitions.RawMonoid` action of a Boolean on a RawMonoid:
+  ```agda
+  _∧_    : Bool → Carrier → Carrier
````
Could be something like `?>` by analogy to `$>`.

gallais replied:
`$>` is `Functor m => m a -> b -> m b`. `<` and `>` always point to the value that is kept in the result.

jamesmckinna commented on src/Algebra/Properties/Monoid/Mult.agda:
```diff
@@ -59,10 +61,7 @@ open import Algebra.Definitions.RawMonoid rawMonoid public
-  x + (m ℕ.+ n) × x    ≈⟨ +-cong refl (×-homo-+ x m n) ⟩
-  x + (m × x + n × x)  ≈⟨ sym (+-assoc x (m × x) (n × x)) ⟩
-  x + m × x + n × x    ∎
+×-homo-+ x (suc m) n = sym (uv≈w⇒xu∙v≈xw (sym (×-homo-+ x m n)) x)
```
Refactored following @JacquesCarette 's suggestion.

jamesmckinna commented:
Converting back from `DRAFT`:
* fixed proofs (incl. @JacquesCarette 's suggestions)
* fixed notation to be (more like) @gallais 's suggestion
* fixed merge conflicts to resolve `CHANGELOG` problems etc.
